feat(tensor): support external buffer when creating input/output tensors

gdh1995 opened the pull request:
**Intent**: Up to now TIM-VX allocates necessary tensor buffer on host memory, and `vsi_nn_AddTensorFromHandle` does accept a non-null `data` argument, so this PR enables a new usage of reusing data buffers from outsides for input and output tensors.

This PR extends and replaces #297.

sunshinemyson commented:
@gdh1995 ,

Are you forcing application provide data for Input/Output?

gdh1995 commented:
Yes. In my code input images come from other threads and I want to reduce counts of copying operations.

If without passing external buffer during constructing graphs, the current process is:
1. other modules generates a NCHW data block `src_data` in thread A
2. thread B calls `input_tensor->CopyDataToTensor(const void* src_data, ...)` to prepare input
3. `CopyDataToTensor` will call `memcpy` when `is_created_from_handle` is `TRUE`
4. when accessing output, `output_tensor->CopyDataFromTensor` means another `memcpy`

With this PR applied in my local clone, I just need to add two small methods:
1. one is for inputting and calls `vsi_nn_FlushHandle(const vsi_nn_tensor_t* input_tensor_inner_ptr)` when input is ready
2. the other is for outputting and calls `vsi_nn_GetTensorHandle` before accessing output content
3. these seems not so necessary - I just want to simulate the official process (but without `memcpy`) as far as possible
4. it works well eduring my tests, which includes some real-time object detection tasks on a video camera.

sunshinemyson commented:
@gdh1995 ,

I agree with you about the API usage. But I didn't see the new API in the commit instead of API behavior change - by replace null_ptr with an application buffer.

Where can i find out the full patch ?

gdh1995 commented:
Um, I didn't add my wrapper of internal APIs like `vsi_nn_FlushHandle`, because I'm not sure how and where to do so. I can provide a naive version of my APIs here, though of course it's not like what a formal API should be.

### My current hack

``` cpp
// TimvxHack.cpp
#include "TimvxHack.h"

#define private public
#define protected public
#include "graph_private.h"
#undef private
#undef protected

#include "tim/vx/direct_map_op.h"
#include "direct_map_op_impl.h"
#include "vsi_nn_tensor_util.h"

using namespace TimvxHack;

namespace TimvxHack {

vsi_nn_tensor_hack_t* GetUndergroundTensor(Context *Context, Graph *GraphInst, Tensor *Tensor)
{
  tim::vx::GraphImpl &Graph = *(tim::vx::GraphImpl *)GraphInst;
  if (Graph.context_ != (void *)Context)
  {
    return nullptr;
  }
  auto TensorID = Tensor->GetId();
  auto UndergroundGraph = Graph.graph_;
  auto UndergroundTensor = vsi_nn_GetTensor(UndergroundGraph, TensorID);
  if (UndergroundTensor == nullptr)
  {
    fprintf(stderr, "Can not get a C-style underground tensor instance from TIM-VX!\n");
  }
  return static_cast<vsi_nn_tensor_hack_t*>(static_cast<void *>(UndergroundTensor));
}

bool GetTensorHandle(vsi_nn_tensor_hack_t* tensor, void** ptr)
{
  return vsi_nn_GetTensorHandle(static_cast<vsi_nn_tensor_t *>(static_cast<void*>(tensor)), ptr) == VSI_SUCCESS;
}

bool FlushHandle(vsi_nn_tensor_hack_t* tensor)
{
  return vsi_nn_FlushHandle(static_cast<const vsi_nn_tensor_t*>(static_cast<const void *>(tensor))) == VSI_SUCCESS;
}

void ReleaseOutputData(vsi_nn_tensor_hack_t* tensor)
{
  // TensorImpl::CopyDataFromTensor in TIM-VX/src/tim/vx/tensor.cc:153 doesn't "release a tensor back"
  return;
}

} // namespace TimvxHack
```

and 

``` cpp
// TimvxHack.h
#pragma once

#include <stdio.h>
#include <string>
#include <vector>
#include "tim/vx/context.h"
#include "tim/vx/graph.h"
#include "tim/vx/tensor.h"
#include "tim/vx/operation.h"

namespace TimvxHack {
  using namespace tim::vx;
  struct vsi_nn_tensor_hack_t;

  vsi_nn_tensor_hack_t* GetUndergroundTensor(Context *Context, Graph *GraphInst, Tensor *Tensor);
  bool GetTensorHandle(vsi_nn_tensor_hack_t * tensor, void** ptr);
  bool FlushHandle(vsi_nn_tensor_hack_t * tensor);
  void ReleaseOutputData(vsi_nn_tensor_hack_t * tensor);
}
```

This `TimvxHack.cpp` file is compiled with an extra include directory of `(TIM-VX/)src/tim/vx/internal/include` as well as public include directories, and then `TimvxHack.h` can be used without `src/tim/vx/internal` imported.

### My proposal of a formal API

If converting my hack to a public and formal API, then we may expose 3 new virtual method function to `tim::vx::Tensor`:
``` cpp
// include/tim/vx/tensor.h
namespace tim {
namespace vx {

class Tensor {
  // ...
public:
  virtual bool FlushDataToTensor();
  virtual bool FlushDataFromTensor();
}

// src/tim/vx/tensor_private.h and src/tim/vx/tensor.cc
class TensorImpl : public Tensor {
public:
  bool FlushDataToTensor() {
    // some checks
    if (tensor && tensor->attr.is_created_from_handle) {
      vsi_nn_FlushHandle(tensor);
      return true;
    }
    return false;
  }
  bool FlushDataFromTensor() {
    // some checks
    auto tensor = vsi_nn_GetTensor(graph_, id_);
    if (tensor && tensor->attr.is_created_from_handle) {
      // TODO(gdh1995): maybe this following block can be skipped completely?
      void* ptr = NULL;
      vsi_nn_GetTensorHandle(tensor, &ptr);
      return true;
    }
    return false;
  }
};

}
}
```

sunshinemyson commented:
@gdh1995,

Thanks for sharing. Would you mind refine your hack and make a PR for this?

API naming prefers tim::vx::Tensor::FlushCacheForHandle, tim::vx::Tensor::InvalidateCacheForHandle instead of 
tim::vx::Tensor::FlushDataToTensor/FlushDataFromTensor.

And for the Hack part, i think you cant add two api in tensor too
Tensor::map(bool invalidate_cache = false):void* cpu_ptr // invoke InvalidateCacheFromHandle inside is optional, some times the memory could no cache at all. cpu_ptr could be null_ptr if tensor itself not create from a handle.

Tensor::unmap(): void : empty implementation if handle created from memory. but for dma_buf_fd, we should unmap it explicitly(we can leave it todo).

Thanks

gdh1995 commented:
Thanks for your help and I'll take a try. But I have 2 problems:

Firstly, do `InvalidateCacheFromHandle` and `InvalidateCacheForHandle` mean a same function?


Secondly, I'm confused by the name of `InvalidateCacheForHandle`. It seems it replaces `FlushDataFromTensor`, but when I firstly noticed the name, I thought it would mean:
- 1. A caller has prepared enough data and written it into `TensorImpl::data_`
- 2. Now it wants to tell TIM-VX "hey, data is prepared, and you should begin your work"
- 3. Then TIM-VX will call `vsi_nn_FlushHandle` and sync data from `TensorImpl::data_` to an inner (and maybe physical) buffer

I think this problem comes from the different understanding of the word `cache` in `InvalidateCacheForHandle` and `FlushCacheForHandle`. Is there a perfect way to avoid such a misunderstanding when other users learn them?

sunshinemyson commented:
@gdh1995 ，

 >> Firstly, do InvalidateCacheFromHandle and InvalidateCacheForHandle mean a same function?  

Yes, they are same function. Sorry for typo.

>> When you provide data to NPU by CopyDataToTensor(), you need to flush CPU cache before NPU continue. And Once NPU is done, before CopyTensorFromData or use Map() to have CPU directly, you need to invalid cpu cache before read.

The naming is little bits confusion because user may not understand which CACHE is flushed or invalidated. Maybe we can have a simple unit test and doc to explain how to use those API.

I think its ok if you want to keep your naming.

Thanks

gdh1995 commented:
Hello, I pushed a new commit and added such public APIs:
1. `virtual bool Tensor::FlushCacheForHandle() = 0;`
2. `virtual bool Tensor::RefillCacheFromHandle() = 0;`
3. `virtual void* Tensor::map(bool refill_cache = false) = 0;`
4. `virtual void Tensor::unmap() = 0;`
5. `virtual std::shared_ptr<Tensor> Graph::CreateIOTensor(const TensorSpec& spec, void* data = nullptr) = 0;`

And added:
1. corresponding member functions in `TensorImpl` and `GraphImpl` classes
2. `TensorImpl::TensorImpl(Graph* graph, const TensorSpec& spec, void* data = nullptr);`

I also redefined `TensorImpl::data_` from `const void *` to `void *`,  which is to indicate this may work as a cache area and be updated by `Tensor::RefillCacheFromHandle` and `Tensor::map(true)`.

If there're any inappropriate place, please let me know.

sunshinemyson commented on src/tim/vx/tensor.cc:
```diff
@@ -167,7 +183,79 @@ bool TensorImpl::CopyDataFromTensor(void* data) {
+        cpu_ptr = data_;
+      } else {
+        vsi_nn_GetTensorHandle(tensor, &cpu_ptr);
+        // TODO: what to do when fd_ != -1
```
If the "managed" data is from fd, we can just use mmap() to retrieve the memory pointer.

gdh1995 replied:
Um, because I don't know much about `fd_` or how to create a valid `fd`, I can not implement it correctly. Could you modify this PR directly? Or may we leave it as is? Then you official developers may replace it in following development.

sunshinemyson replied:
Okay, You can just keep it as TODO. We can follow-up later.

gdh1995 commented on include/tim/vx/tensor.h:
```diff
@@ -154,6 +154,10 @@ class Tensor {
   virtual bool CopyDataToTensor(const void* data, uint32_t size_in_bytes = 0) = 0;
   virtual bool CopyDataFromTensor(void* data) = 0;
+  virtual bool FlushCacheForHandle() = 0;
+  virtual bool RefillCacheFromHandle() = 0;
```
I use `RefillCacheFromHandle` on `attr=output` tensors, and in my perspective, this name means: a Tensor instance fills the cpu cache with computed feature maps. I want to give such a promise: after this function returns, the cpu cache contains valid feature map data.

gdh1995 replied:
Or do I misunderstand the `cache`? I guess it means the (external or internal-created) aligned buffer of `data_`. But if its real meaning is a memory block maintained inside the tensor handle, then this difference may explain why I could not get what you said.

sunshinemyson replied:
cache is simple, it is the cache from CPU point of view. We don't have any "cache" inside driver/npu-hw.

gdh1995 replied:
Um, I found https://github.com/repo/VeriSilicon/TIM-VX/issues/143, then maybe I've got what you meant.

gdh1995 replied:
Then the name of `InvalidateCacheForHandle` is perfect. I've updated the code to use this name, and added auto flushing input tensors in `TensorImpl::unmap`.
